pages/developers/gha_basic.md

```diff
@@ -0,0 +1,164 @@
+    - uses: actions/cache@v1
+      with:
+        path: ~/.cache/pre-commit
+        key: pre-commit|${{ env.PY }}|${{ hashFiles('.pre-commit-config.yaml') }}
+    - uses: pre-commit/action@v1.1.0
+```
+
+## Unit tests
+
+Implementing unit tests is also easy. Since you should be following best
+practices listed in the previous sections, this becomes an almost directly
+copy-and-paste formula, regardless of the package details.
+
+```yaml
+  checks:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        python-version:
+        - 2.7
+        - 3.6
+        - 3.8
+    name: Check Python ${{ matrix.python-version }}
+    steps:
+    - uses: actions/checkout@v1
+    - uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+
+    - name: Install package
+      run: python -m pip install -e .[test]
+
+    - name: Test package
+      run: python -m pytest
+```
+
+A few things to note from above:
+
+The matrix should contain the versions you are interested in. You can also test
+on other OS's if you are building any extensions or are worried about your
+package on macOS or Windows. Fail-fast is optional.
+
+You need to use version 1 of the checkout, since version 2 strips too much
+from the repository for `setuptools_scm` to work.
+
+The formula here for installing should be identical for all users; and using
+PEP 517/518 builds, you are even guaranteed a consistent wheel will be produced
+just as if you were building a final package.
+
+## Distribution: Pure Python wheels
+
+We will cover binary wheels later, but for a simple universal (pure Python)
+package, the procedure is simple.
+
+```yaml
+  dist:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v1
+    - uses: actions/setup-python@v1
+      with:
+        python-version: 3.8
+
+    - name: Install wheel and sdist requirements
+      run: python -m pip install "setuptools>=42.0" "setuptools_scm[toml]>=3.4" "wheel"
+
+    - name: Build sdist
+      run: python setup.py sdist
+
+    - name: Build wheel
+      run: >
+        python -m pip wheel . -w wheels &&
+        ls -lh wheels &&
+        mkdir -p dist &&
+        cp wheels/<packagename>*any.whl dist/
+
+    - uses: actions/upload-artifact@v1
+      with:
+        name: DistPackage
+        path: dist
+
+    - uses: pypa/gh-action-pypi-publish@master
+      with:
+        user: __token__
+        password: ${{ secrets.pypi_password }}
+      if: github.event_name == 'push' && startsWith(github.event.ref, 'refs/tags')
+```
+
+A few things to note that are new to this job:
+
+We install SDist requirements by hand since `python setup.py sdist` does not
+get the benefits of having pip install things. If you have any special
+requirements in your `pyproject.toml`, you'll need to list them here. This is
+special just for the sdist, not for making wheels (which should be done by the
```

Comment:
Rather type "SDist" for consistency with 3 lines above.

Reply:
:blush: